File: Java/src/main/java/Solution601_700/Solution645.java
package Solution601_700;

import java.util.Arrays;

/**
 * @author windows
 */
public class Solution645 {
    public int[] findErrorNums(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1){
                int temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
                if (nums[i] != nums[nums[i] - 1]){
                    i--;
                }
            }
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1){
                return new int[]{nums[i], i + 1};
            }
        }
        return new int[2];
    }


    public static void main(String[] args) {
        Solution645 solution645 = new Solution645();
        System.out.println(Arrays.toString(solution645.findErrorNums(new int[]{2, 3, 2})));
    }
}

File: C++/src/Solution601_700/Solution645.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution645.h"
#include <algorithm>

vector<int> Solution645::findErrorNums(vector<int> &nums) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<int> res(2);
    int prev = 0;
    for (int i = 0; i < n; ++i) {
        int cur = nums[i];
        if (cur == prev){
            res[0] = cur;
        } else if (cur - prev > 1){
            res[1] = prev + 1;
        }
        prev = cur;
    }
    if (nums[n - 1] != n){
        res[1] = n;
    }
    return res;
}
